kolibri/content/utils/channel_import.py

```diff
@@ -0,0 +1,287 @@
+
+
+class NoVersionChannelImport(ChannelImport):
+    """
+    Class defining the schema mapping for importing old content databases (i.e. ones produced before the
+    ChannelImport machinery was implemented). The schema mapping below defines how to bring in information
+    from the old version of the Kolibri content databases into the database for the current version of Kolibri.
+    """
+
+    schema_mapping = {
+        ContentNode: {
+            'per_row': {
+                'channel_id': 'infer_channel_id_from_source',
+                'tree_id': 'get_tree_id',
+                'available': 'none',
+            },
+        },
+        File: {
+            'per_row': {
+                File._meta.get_field('local_file').attname: 'checksum',
+                'available': 'none',
+            },
+        },
+        LocalFile: {
+            'per_table': 'generate_local_file_from_file',
+            'per_row': {
+                'id': 'checksum',
+                'extension': 'extension',
+                'file_size': 'file_size',
+                'available': 'none',
+            },
+        },
+        ChannelMetadata: {
+            'per_row': {
+                ChannelMetadata._meta.get_field('min_kolibri_version').attname: 'set_version_to_no_version',
+                'root_id': 'root_pk',
+            },
+        },
+    }
+
+    def none(self, source_object):
+        return None
+
+    def infer_channel_id_from_source(self, source_object):
+        return self.channel_id
+
+    def get_tree_id(self, source_object):
+        return self.tree_id
+
+    def generate_local_file_from_file(self, SourceRecord):
+        SourceRecord = self.source.get_class(File)
+        checksum_record = set()
+        # LocalFile objects are unique per checksum
+        for record in self.source.session.query(SourceRecord).all():
+            if record.checksum not in checksum_record:
+                checksum_record.add(record.checksum)
+                yield record
+            else:
+                continue
+
+    def set_version_to_no_version(self, source_object):
+        return NO_VERSION
+
+
+mappings = {
+    NO_VERSION: NoVersionChannelImport,
+    '0.6.0': ChannelImport,
+}
+
+
+def initialize_import_manager(channel_id):
+
+    channel_metadata = read_channel_metadata_from_db_file(get_content_database_file_path(channel_id))
+
+    min_version = getattr(channel_metadata, 'min_kolibri_version', NO_VERSION)
+
+    ImportClass = mappings.get(min_version)
+
+    return ImportClass(channel_id)
+
+
+def import_channel_from_local_db(channel_id):
+    import_manager = initialize_import_manager(channel_id)
+
+    if ChannelMetadata.objects.filter(id=channel_id).exists():
+        # We have already imported this channel in some way, so let's clean up first.
+        logging.info('Channel {channel_id} already exists in database, cleaning up ContentNodes'.format(channel_id=channel_id))
+        import_manager.delete_content_tree_and_files()
```

Comment:
Because I am using the Django ORM for this delete, it happens in a separate transaction (otherwise I don't get the cascade deletion behaviour). With a bit more work, I think I could replicate the django cascade deletion in SQLAlchemy, however, and then we could wrap everything in a transaction.